Prediction Guard Guardrails components

sharanshirodkar7 opened the pull request:
## Description

This PR introduces new guardrail components for Prediction Guard (PG) to enhance the safety and reliability of AI interactions. The following guardrails have been added:

	1.	Factuality Detection
	2.	Toxicity Detection
	3.	Prompt Injection Detection
	4.	PII detection

## Issues

n/a

## Type of change

New feature (non-breaking change which adds new functionality)

## Dependencies

PREDICTIONGUARD_API_KEY needs to be added to CI/CD to test the microservice.

## Tests

The following tests were conducted to verify the functionality and integration of the new guardrail components:

	•	Unit tests for each guardrail component to validate detection accuracy and reliability.
	•	Integration tests to ensure smooth operation alongside existing Prediction Guard components.
	•	Performance tests to verify the guardrails do not introduce significant latency or performance overhead.

All tests have passed, confirming the effective integration of the guardrail components.

sharanshirodkar7 commented:
@chensuyue PG guardrails comp ready for review.

lvliang-intel commented:
@sharanshirodkar7,

Please fix the pre-commit issue.

![image](https://github.com/user-attachments/assets/f4201352-9226-42f3-9c85-26792a0c947d)

sharanshirodkar7 commented:
@lvliang-intel updated.

sharanshirodkar7 commented:
@letonghan  @lvliang-intel ready for review.

chensuyue commented:
What's the relationship between those 2 dockerfile, if they belong to 2 different implementation for the service, they the original one should be package into a folder. Based on the new structure, the path with `Dockerfile*` will be recognized as the service path belong to single service. That means when the CI recognize `comps/guardrails/pii_detection/predictionguard` belong to `comps/guardrails/pii_detection`. I think that not what you want. 
![image](https://github.com/user-attachments/assets/1bb9572a-87c5-4e3c-89e8-7b5bf92dbd5e)
![image](https://github.com/user-attachments/assets/fe46cabd-cb02-4373-b2d6-4b0a4639dcc6)

sharanshirodkar7 commented:
@chensuyue so those files already exist in the repo main right now outside of any folder within toxicity detection. I created a folder predictionguard and added our files there. I was not sure if I have to move the files not related to our service as it might affect some links in examples repo too. Let me know what we should do here.

sharanshirodkar7 commented:
@letonghan ready for review and merge.

letonghan commented on comps/guardrails/factuality/predictionguard/factuality_predictionguard.py:
```diff
@@ -0,0 +1,48 @@
+from comps import ServiceType, opea_microservices, register_microservice, register_statistics, statistics_dict
+
+
+class FactualityDoc(BaseDoc):
```
These new Doc should be moved to `GenAIComps\comps\cores\proto\docarray.py`

letonghan commented on comps/guardrails/factuality/predictionguard/requirements.txt:
```diff
@@ -0,0 +1,13 @@
+opentelemetry-exporter-otlp
+opentelemetry-sdk
+Pillow
+predictionguard==2.2.1
```
why use `predictionguard==2.2.1`? this could cause unmatched dependency version when other dependencies update.

letonghan commented on comps/guardrails/pii_detection/predictionguard/pii_predictionguard.py:
```diff
@@ -0,0 +1,57 @@
+from comps import ServiceType, opea_microservices, register_microservice, register_statistics, statistics_dict
+
+
+class PIIRequestDoc(BaseDoc):
```
same here, need to move to `GenAIComps\comps\cores\proto\docarray.py`

letonghan commented:
Can you explain the relationship of `prediction guard` and other four implementations (Factuality Detection, Toxicity Detection, ...)? So we can discuss how to reorg your new components folders.
After that, remember to update the README of `GenAIComps\comps\guardrails\README.md`. Thanks : )

sharanshirodkar7 commented:
@lvliang-intel, thanks for your feedback. I’m working on addressing your suggestions.
@letonghan, sure so Prediction Guard is an end-to-end GenAI platform that provides access to the latest open models from Hugging Face, including LLMs, LVMs, and embeddings (these three microservices have already been merged into OPEA). The platform’s key value proposition is its focus on privacy and security features, such as PII detection, prompt injection checks, factuality, and toxicity checks. These services are integral to Prediction Guard, which is why I created dedicated folders for predictionguard and added the corresponding code.

sharanshirodkar7 commented:
@lvliang-intel I updated the codes with the changes you mentioned- cd ../../../.. , readme's and requirements versioning. Now it shows some new checks to be failing readme hyperlinks and compose yaml references. Let me know if I need to take any action here.

sharanshirodkar7 commented:
@lvliang-intel updated.
